eclair-core/src/main/scala/fr/acinq/eclair/wire/protocol/PaymentOnion.scala

```diff
@@ -185,91 +185,109 @@ object PaymentOnion {
+    val cltvOut = records.get[OutgoingCltv].get.cltv
     override val outgoingChannelId = records.get[OutgoingChannelId].get.shortChannelId
+    override def amountToForward(incomingAmount: MilliSatoshi): MilliSatoshi = amountOut
+    override def outgoingCltv(incomingCltv: CltvExpiry): CltvExpiry = cltvOut
+    // @formatter:on
   }
 
-  object ChannelRelayTlvPayload {
-    def apply(outgoingChannelId: ShortChannelId, amountToForward: MilliSatoshi, outgoingCltv: CltvExpiry): ChannelRelayTlvPayload =
-      ChannelRelayTlvPayload(TlvStream(OnionPaymentPayloadTlv.AmountToForward(amountToForward), OnionPaymentPayloadTlv.OutgoingCltv(outgoingCltv), OnionPaymentPayloadTlv.OutgoingChannelId(outgoingChannelId)))
+  object ChannelRelayPayload {
+    def apply(outgoingChannelId: ShortChannelId, amountToForward: MilliSatoshi, outgoingCltv: CltvExpiry): ChannelRelayPayload = {
+      ChannelRelayPayload(TlvStream(OnionPaymentPayloadTlv.AmountToForward(amountToForward), OnionPaymentPayloadTlv.OutgoingCltv(outgoingCltv), OnionPaymentPayloadTlv.OutgoingChannelId(outgoingChannelId)))
+    }
+
+    def validate(records: TlvStream[OnionPaymentPayloadTlv]): Either[InvalidTlvPayload, ChannelRelayPayload] = {
+      if (records.get[AmountToForward].isEmpty) return Left(MissingRequiredTlv(UInt64(2)))
+      if (records.get[OutgoingCltv].isEmpty) return Left(MissingRequiredTlv(UInt64(4)))
+      if (records.get[OutgoingChannelId].isEmpty) return Left(MissingRequiredTlv(UInt64(6)))
+      Right(ChannelRelayPayload(records))
+    }
   }
 
-  case class BlindedChannelRelayPayload(records: TlvStream[OnionPaymentPayloadTlv]) extends ChannelRelayPayload {
-    val blinding_opt: Option[PublicKey] = records.get[BlindingPoint].map(_.publicKey)
-    val encryptedRecipientData: ByteVector = records.get[EncryptedRecipientData].get.data
+  /**
+   * @param blindedRecords decrypted tlv stream from the encrypted_recipient_data tlv.
+   * @param nextBlinding   blinding point that must be forwarded to the next hop.
+   */
+  case class BlindedChannelRelayPayload(records: TlvStream[OnionPaymentPayloadTlv], blindedRecords: TlvStream[RouteBlindingEncryptedDataTlv], nextBlinding: PublicKey) extends ChannelRelayHopPayload {
+    // @formatter:off
+    override val outgoingChannelId = blindedRecords.get[RouteBlindingEncryptedDataTlv.OutgoingChannelId].get.shortChannelId
+    val paymentRelay = blindedRecords.get[RouteBlindingEncryptedDataTlv.PaymentRelay].get
+    val paymentConstraints = blindedRecords.get[RouteBlindingEncryptedDataTlv.PaymentConstraints].get
+    val allowedFeatures = blindedRecords.get[RouteBlindingEncryptedDataTlv.AllowedFeatures].map(_.features).getOrElse(Features.empty)
+    override def amountToForward(incomingAmount: MilliSatoshi): MilliSatoshi = ((incomingAmount - paymentRelay.feeBase).toLong * 1_000_000 + 1_000_000 + paymentRelay.feeProportionalMillionths - 1).msat / (1_000_000 + paymentRelay.feeProportionalMillionths)
```

Comment:
It's just standard formula to avoid truncation with integer arithmetic, see https://github.com/ACINQ/eclair/pull/2367
Rusty mentioned that this is how `ceil` operations are done inside the linux kernel.